hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/extensions/MultiNetworkExtension.java

```diff
@@ -0,0 +1,172 @@
+                                    networks.length));
+        });
+    }
+
+    @Override
+    public void afterEach(@NonNull final ExtensionContext extensionContext) {
+        final var networks = store(extensionContext).remove(RESOURCES_KEY, SubProcessNetwork[].class);
+        if (networks != null) {
+            for (final var network : networks) {
+                safeTerminate(network);
+            }
+        }
+        store(extensionContext).remove(PARAM_INDEXES_KEY);
+    }
+
+    @Override
+    public boolean supportsParameter(
+            @NonNull final ParameterContext parameterContext, @NonNull final ExtensionContext extensionContext) {
+        final var type = parameterContext.getParameter().getType();
+        final boolean isSupportedType =
+                HederaNetwork.class.isAssignableFrom(type) || SubProcessNetwork.class.isAssignableFrom(type);
+        return isSupportedType && findAnnotation(extensionContext).isPresent();
+    }
+
+    @Override
+    public Object resolveParameter(
+            @NonNull final ParameterContext parameterContext, @NonNull final ExtensionContext extensionContext) {
+        final var networks = store(extensionContext).get(RESOURCES_KEY, SubProcessNetwork[].class);
+        final var indexes = store(extensionContext).get(PARAM_INDEXES_KEY, List.class);
+        if (networks == null || indexes == null) {
+            throw new IllegalStateException("Multi networks have not been initialized for this test");
+        }
+        final var index = parameterContext.getIndex();
+        final var networkPosition = indexes.indexOf(index);
+        if (networkPosition == -1) {
+            throw new IllegalArgumentException("Parameter at index " + index + " is not mapped to a network");
+        }
+        if (networkPosition >= networks.length) {
+            throw new IllegalStateException("Unexpected network parameter position: " + networkPosition);
+        }
+        return networks[networkPosition];
+    }
+
+    private SubProcessNetwork[] startNetworks(@NonNull final Network[] networkConfigs) {
+        if (networkConfigs.length == 0) {
+            throw new IllegalStateException("MultiNetworkHapiTest requires at least one network");
+        }
+        final Map<String, Integer> nameCounts = Arrays.stream(networkConfigs)
+                .map(Network::name)
+                .collect(Collectors.toMap(Function.identity(), name -> 1, Integer::sum));
+        nameCounts.forEach((name, count) -> {
+            if (count > 1) {
+                throw new IllegalArgumentException("Network names must be unique, found duplicate: " + name);
+            }
+        });
+
+        final List<SubProcessNetwork> networks = new ArrayList<>();
+        final long shard = getConfigShard();
+        final long realm = getConfigRealm();
+        for (final var config : networkConfigs) {
+            final var network = SubProcessNetwork.newIsolatedNetwork(config.name(), config.size(), shard, realm);
```

Comment:
Why does this pass in the same shard/realm for every network here? That seems incorrect

Reply:
I _think_ so. Still a bit fuzzy (see earlier comment). This may somewhat encourage incorrect assumptions about shard/realm, but not concerned for the prototype